stock_putaway_product_form/models/product_template.py

```diff
@@ -0,0 +1,62 @@
+# License AGPL-3.0 or later (http://www.gnu.org/licenses/agpl).
+
+from odoo import api, fields, models
+
+
+class ProductTemplate(models.Model):
+    _inherit = "product.template"
+
+    product_tmpl_putaway_ids = fields.One2many(
+        comodel_name="stock.fixed.putaway.strat",
+        inverse_name="product_tmpl_id",
+        string="Product putaway strategies by product",
+    )
+
+    product_putaway_categ_ids = fields.Many2many(
+        comodel_name="stock.fixed.putaway.strat",
+        string="Product putaway strategies by category",
+        compute="_compute_putaway_categ_ids",
+    )
+
+    def _find_closest_categ_match(self, categ, putaway_lines):
+        """Returns the putaway line with the nearest product category"""
+        lines_match_categ = putaway_lines.filtered(
+            lambda r: r.category_id == categ
+        )
+        if lines_match_categ:
+            return lines_match_categ[0]
+        elif categ.parent_id:
+            return self._find_closest_categ_match(
+                categ.parent_id, putaway_lines
+            )
+        else:
+            return self.env["stock.fixed.putaway.strat"]
+
+    def _get_categ_and_parents(self, categ):
+        parent_categ_iterator = categ
+        res = self.env["product.category"]
+        while parent_categ_iterator:
+            res += parent_categ_iterator
+            parent_categ_iterator = parent_categ_iterator.parent_id
+        return res
+
+    @api.depends("categ_id")
+    def _compute_putaway_categ_ids(self):
+        """Pay attention to keep only 1 (most specific,
+        i.e closest to our product category's parents)
+        putaway.strat per product.putaway"""
+        for rec in self:
+            res = self.env["stock.fixed.putaway.strat"]
+            categ = rec.categ_id
+            categs = self._get_categ_and_parents(categ)
+            # get matching lines from our category or its parents
+            product_putaway_categ_lines = self.env[
```

Comment:
I think the following code is broken.
Indeed in categs you have all parent categs
In product_putaway_categ_lines we will have all putaway for all this categs but the order will not respect the order of the category ! This mean you can take the strategy of the grand parent instead of the parent.

Reply:
you can reuse the same code https://github.com/OCA/OCB/blob/12.0/addons/stock/models/product_strategy.py#L41

Reply:
I use the same reasoning as the while loop you linked, except it is in the form of the recursive function _find_closest_categ_match (they look a bit different but the end result is similar).

We start from all ancestors without caring about order because we first need to find the putaway lines (stock.fixed.putaway.strat) that might be relevant, then mapping their parent (putaway_id), because we only want one closest parent per product.putaway (so we operate on product.putaway basis instead of a putaway line basis). Then, for each putaway strategy, call the function that is equivalent to that while loop. So ancestor order is taken into account.

I don't see a more efficient solution. If you think the while loop is clearer than the _find_closest_categ_match function I can use it instead, but the reasoning will be the same